lib/pbio/drv/usb/stm32_usbd/usbd_desc.c

```diff
@@ -231,17 +386,81 @@ uint8_t *USBD_VCP_ConfigStrDescriptor(USBD_SpeedTypeDef speed, uint16_t *length)
   * @param  length: Pointer to data length variable
   * @retval Pointer to descriptor buffer
   */
-uint8_t *USBD_VCP_InterfaceStrDescriptor(USBD_SpeedTypeDef speed, uint16_t *length) {
+static uint8_t *USBD_Pybricks_InterfaceStrDescriptor(USBD_SpeedTypeDef speed, uint16_t *length) {
     USBD_GetString((uint8_t *)USBD_INTERFACE_FS_STRING, USBD_StrDesc, length);
     return USBD_StrDesc;
 }
 
-USBD_DescriptorsTypeDef VCP_Desc = {
-    USBD_VCP_DeviceDescriptor,
-    USBD_VCP_LangIDStrDescriptor,
-    USBD_VCP_ManufacturerStrDescriptor,
-    USBD_VCP_ProductStrDescriptor,
-    USBD_VCP_SerialStrDescriptor,
-    USBD_VCP_ConfigStrDescriptor,
-    USBD_VCP_InterfaceStrDescriptor,
+#if ((USBD_LPM_ENABLED == 1) || (USBD_CLASS_BOS_ENABLED == 1))
+static uint8_t *USBD_Pybricks_BOSDescriptor(USBD_SpeedTypeDef speed, uint16_t *length) {
+    /* Prevent unused argument(s) compilation warning */
+    UNUSED(speed);
+
+    static uint8_t created = 0;
+    uint8_t *ptr;
+
+    /* Generate BOS Descriptor on first attempt */
+    if (!created) {
```

Comment:
We can, but I think this complicates the interface. It would be nice if `USBD_DescriptorsTypeDef` included an `Init` like `USBD_ClassTypeDef` does.

Reply:
It's not pretty, but I think it can be done. Is it worth it for saving 115 bytes of RAM? :shrug: 

```diff
diff --cc lib/pbio/drv/usb/stm32_usbd/usbd_desc.c
index 4a28709e,4a28709e..20e4f5e7
--- a/lib/pbio/drv/usb/stm32_usbd/usbd_desc.c
+++ b/lib/pbio/drv/usb/stm32_usbd/usbd_desc.c
@@@ -49,6 -49,6 +49,7 @@@
  
  #include <pbdrv/config.h>
  #include <pbio/protocol.h>
++#include <pbio/version.h>
  
  #include "usbd_core.h"
  #include "usbd_conf.h"
@@@ -67,21 -67,21 +68,23 @@@
  #define USBD_CONFIGURATION_FS_STRING  "Pybricks Config"
  #define USBD_INTERFACE_FS_STRING      "Pybricks Interface"
  
--static const char firmware_version[] = PBIO_VERSION_STR;
--static const char software_version[] = PBIO_PROTOCOL_VERSION_STR;
--
  #define         DEVICE_ID1          (0x1FFF7A10)
  #define         DEVICE_ID2          (0x1FFF7A14)
  #define         DEVICE_ID3          (0x1FFF7A18)
  
  #define  USB_SIZ_STRING_SERIAL       0x1A
--#define  USB_SIZ_BOS_DESC_CONST      (5 + 28)
--#define  USB_SIZ_UUID                (128 / 8)
--#define  USB_SIZ_PLATFORM_HDR        (4 + USB_SIZ_UUID)
--#define  USB_SIZ_BOS_DESC            (USB_SIZ_BOS_DESC_CONST + \
--    USB_SIZ_PLATFORM_HDR + sizeof(firmware_version) + \
--    USB_SIZ_PLATFORM_HDR + sizeof(software_version) + \
--    USB_SIZ_PLATFORM_HDR + PBIO_PYBRICKS_HUB_CAPABILITIES_VALUE_SIZE)
++#define  USB_SIZ_BOS_DESC (5 + \
++    28 + \
++    20 + sizeof(PBIO_VERSION_STR) - 1 + \
++    20 + sizeof(PBIO_PROTOCOL_VERSION_STR) - 1 + \
++    20 + PBIO_PYBRICKS_HUB_CAPABILITIES_VALUE_SIZE)
++
++/** Rearrange UUID bytes to match RFC 4122 format. */
++#define RFC_4122_UUID(b0, b1, b2, b3, b4, b5, b6, b7, b8, b9, ba, bb, bc, bd, be, bf) \
++    b3, b2, b1, b0, \
++    b5, b4, \
++    b7, b6, \
++    b8, b9, ba, bb, bc, bd, be, bf
  
  /* USB Standard Device Descriptor */
  __ALIGN_BEGIN static
@@@ -111,36 -111,36 +114,100 @@@ uint8_t USBD_DeviceDesc[USB_LEN_DEV_DES
  }; /* USB_DeviceDescriptor */
  
  /** BOS descriptor. */
--__ALIGN_BEGIN static uint8_t USBD_BOSDesc[USB_SIZ_BOS_DESC] __ALIGN_END =
++__ALIGN_BEGIN static const uint8_t USBD_BOSDesc[USB_SIZ_BOS_DESC] __ALIGN_END =
  {
--    5,                        /* bLength */
--    USB_DESC_TYPE_BOS,        /* bDescriptorType = BOS */
--    LOBYTE(USB_SIZ_BOS_DESC), /* wTotalLength */
--    HIBYTE(USB_SIZ_BOS_DESC), /* wTotalLength */
--    0x01,                     /* bNumDeviceCaps = 1 */
--
--    28,                       /* bLength */
--    USB_DEVICE_CAPABITY_TYPE, /* bDescriptorType = Device Capability */
--    0x05,                     /* bDevCapabilityType = Platform */
--    0x00,                     /* bReserved */
--
++    5,                          /* bLength */
++    USB_DESC_TYPE_BOS,          /* bDescriptorType = BOS */
++    LOBYTE(USB_SIZ_BOS_DESC),   /* wTotalLength */
++    HIBYTE(USB_SIZ_BOS_DESC),   /* wTotalLength */
++    0x04,                       /* bNumDeviceCaps */
++
++    28,                             /* bLength */
++    USB_DEVICE_CAPABITY_TYPE,       /* bDescriptorType = Device Capability */
++    0x05,                           /* bDevCapabilityType = Platform */
++    0x00,                           /* bReserved */
      /*
       * PlatformCapabilityUUID
       * Microsoft OS 2.0 descriptor platform capability ID
       * D8DD60DF-4589-4CC7-9CD2-659D9E648A9F
       * RFC 4122 explains the correct byte ordering
       */
--    0xDF, 0x60, 0xDD, 0xD8,           /* 32-bit value */
--    0x89, 0x45,                       /* 16-bit value */
--    0xC7, 0x4C,                       /* 16-bit value */
--    0x9C, 0xD2,
--    0x65, 0x9D, 0x9E, 0x64, 0x8A, 0x9F,
++    RFC_4122_UUID(0xD8, 0xDD, 0x60, 0xDF, 0x45, 0x89, 0x4C, 0xC7,
++        0x9C, 0xD2, 0x65, 0x9D, 0x9E, 0x64, 0x8A, 0x9F),
  
--    0x00, 0x00, 0x03, 0x06,           /* dwWindowsVersion = 0x06030000 for Windows 8.1 Build */
++    0x00, 0x00, 0x03, 0x06,             /* dwWindowsVersion = 0x06030000 for Windows 8.1 Build */
      LOBYTE(USBD_SIZ_MS_OS_DSCRPTR_SET), /* wMSOSDescriptorSetTotalLength */
      HIBYTE(USBD_SIZ_MS_OS_DSCRPTR_SET), /* wMSOSDescriptorSetTotalLength */
--    USBD_MS_VENDOR_CODE,              /* bMS_VendorCode */
--    0x00                              /* bAltEnumCode = Does not support alternate enumeration */
++    USBD_MS_VENDOR_CODE,                /* bMS_VendorCode */
++    0x00,                               /* bAltEnumCode = Does not support alternate enumeration */
++
++    20 + sizeof(PBIO_VERSION_STR) - 1,  /* bLength*/
++    USB_DEVICE_CAPABITY_TYPE,           /* bDescriptorType = Device Capability */
++    0x05,                               /* bDevCapabilityType = Platform */
++    0x00,                               /* bReserved */
++    /* PlatformCapabilityUUID - BLE firmware revision characteristic */
++    RFC_4122_UUID(0x00, 0x00, 0x2A, 0x26, 0x00, 0x00, 0x10, 0x00,
++        0x80, 0x00, 0x00, 0x80, 0x5F, 0x9B, 0x34, 0xFB),
++    // not a nice way to use PBIO_VERSION_STR here but should end up with the
++    // same result
++    #if PBIO_VERSION_MAJOR >= 10
++    '0' + PBIO_VERSION_MAJOR / 10,
++    #endif
++    '0' + PBIO_VERSION_MAJOR % 10,
++    '.',
++    #if PBIO_VERSION_MINOR >= 10
++    '0' + PBIO_VERSION_MINOR / 10,
++    #endif
++    '0' + PBIO_VERSION_MINOR % 10,
++    '.',
++    #if PBIO_VERSION_MICRO >= 10
++    '0' + PBIO_VERSION_MICRO / 10,
++    #endif
++    '0' + PBIO_VERSION_MICRO % 10,
++    #if PBIO_VERSION_LEVEL_HEX != 0xF
++    'a' + PBIO_VERSION_LEVEL_HEX - 0xA,
++    #if PBIO_VERSION_SERIAL >= 10
++    '0' + PBIO_VERSION_SERIAL / 10,
++    #endif
++    '0' + PBIO_VERSION_SERIAL % 10,
++    #endif
++
++    20 + sizeof(PBIO_PROTOCOL_VERSION_STR) - 1, /* bLength*/
++    USB_DEVICE_CAPABITY_TYPE,           /* bDescriptorType = Device Capability */
++    0x05,                               /* bDevCapabilityType = Platform */
++    0x00,                               /* bReserved */
++    /* PlatformCapabilityUUID - BLE software revision characteristic */
++    RFC_4122_UUID(0x00, 0x00, 0x2A, 0x28, 0x00, 0x00, 0x10, 0x00,
++        0x80, 0x00, 0x00, 0x80, 0x5F, 0x9B, 0x34, 0xFB),
++    // not a nice way to use PBIO_PROTOCOL_VERSION_STR here but should end up
++    // with the same result
++    #if PBIO_PROTOCOL_VERSION_MAJOR >= 10
++    '0' + PBIO_PROTOCOL_VERSION_MAJOR / 10,
++    #endif
++    '0' + PBIO_PROTOCOL_VERSION_MAJOR % 10,
++    '.',
++    #if PBIO_PROTOCOL_VERSION_MINOR >= 10
++    '0' + PBIO_PROTOCOL_VERSION_MINOR / 10,
++    #endif
++    '0' + PBIO_PROTOCOL_VERSION_MINOR % 10,
++    '.',
++    #if PBIO_PROTOCOL_VERSION_PATCH >= 10
++    '0' + PBIO_PROTOCOL_VERSION_PATCH / 10,
++    #endif
++    '0' + PBIO_PROTOCOL_VERSION_PATCH % 10,
++
++    20 + sizeof(PBIO_PYBRICKS_HUB_CAPABILITIES_VALUE_SIZE), /* bLength*/
++    USB_DEVICE_CAPABITY_TYPE,           /* bDescriptorType = Device Capability */
++    0x05,                               /* bDevCapabilityType = Platform */
++    0x00,                               /* bReserved */
++    /* PlatformCapabilityUUID - BLE Pybricks hub capability characteristic */
++    RFC_4122_UUID(0xC5, 0xF5, 0x00, 0x03, 0x82, 0x80, 0x46, 0xDA,
++        0x89, 0xF4, 0x6D, 0x80, 0x51, 0xE4, 0xAE, 0xEF),
++    USBD_PYBRICKS_MAX_PACKET_SIZE & 0xFF, (USBD_PYBRICKS_MAX_PACKET_SIZE >> 8) & 0xFF, /* 16-bit LE */
++    PBSYS_APP_HUB_FEATURE_FLAGS & 0xFF, (PBSYS_APP_HUB_FEATURE_FLAGS >> 8) & 0xFF, /* 32-bit LE */
++    (PBSYS_APP_HUB_FEATURE_FLAGS >> 16) & 0xFF, (PBSYS_APP_HUB_FEATURE_FLAGS >> 24) & 0xFF,
++    PBSYS_PROGRAM_LOAD_MAX_PROGRAM_SIZE & 0xFF, (PBSYS_PROGRAM_LOAD_MAX_PROGRAM_SIZE >> 8) & 0xFF, /* 32-bit LE */
++    (PBSYS_PROGRAM_LOAD_MAX_PROGRAM_SIZE >> 16) & 0xFF, (PBSYS_PROGRAM_LOAD_MAX_PROGRAM_SIZE >> 24) & 0xFF,
  };
  
  __ALIGN_BEGIN uint8_t USBD_OSDescSet[USBD_SIZ_MS_OS_DSCRPTR_SET] __ALIGN_END =
@@@ -399,57 -399,57 +466,6 @@@ static uint8_t *USBD_Pybricks_BOSDescri
      /* Prevent unused argument(s) compilation warning */
      UNUSED(speed);
  
--    static uint8_t created = 0;
--    uint8_t *ptr;
--
--    /* Generate BOS Descriptor on first attempt */
--    if (!created) {
--        created = 1;
--        ptr = &USBD_BOSDesc[USB_SIZ_BOS_DESC_CONST];
--
--        /* Add firmware version */
--        *ptr++ = USB_SIZ_PLATFORM_HDR + sizeof(firmware_version);
--        *ptr++ = USB_DEVICE_CAPABITY_TYPE;
--        *ptr++ = 0x05;
--        *ptr++ = 0x00;
--
--        pbio_uuid128_le_copy(ptr, pbio_gatt_firmware_version_char_uuid_128);
--        ptr += USB_SIZ_UUID;
--
--        memcpy(ptr, firmware_version, sizeof(firmware_version));
--        ptr += sizeof(firmware_version);
--
--        /* Add software (protocol) version */
--        *ptr++ = USB_SIZ_PLATFORM_HDR + sizeof(software_version);
--        *ptr++ = USB_DEVICE_CAPABITY_TYPE;
--        *ptr++ = 0x05;
--        *ptr++ = 0x00;
--
--        pbio_uuid128_le_copy(ptr, pbio_gatt_software_version_char_uuid_128);
--        ptr += USB_SIZ_UUID;
--
--        memcpy(ptr, software_version, sizeof(software_version));
--        ptr += sizeof(software_version);
--
--        /* Add hub capabilities */
--        *ptr++ = USB_SIZ_PLATFORM_HDR + PBIO_PYBRICKS_HUB_CAPABILITIES_VALUE_SIZE;
--        *ptr++ = USB_DEVICE_CAPABITY_TYPE;
--        *ptr++ = 0x05;
--        *ptr++ = 0x00;
--
--        pbio_uuid128_le_copy(ptr, pbio_pybricks_hub_capabilities_char_uuid);
--        ptr += USB_SIZ_UUID;
--
--        pbio_pybricks_hub_capabilities(ptr,
--            USBD_PYBRICKS_MAX_PACKET_SIZE - USB_SIZ_UUID,
--            PBSYS_APP_HUB_FEATURE_FLAGS,
--            PBSYS_PROGRAM_LOAD_MAX_PROGRAM_SIZE);
--        ptr += PBIO_PYBRICKS_HUB_CAPABILITIES_VALUE_SIZE;
--
--        /* Update bNumDeviceCaps field in BOS Descriptor */
--        USBD_BOSDesc[4] += 3;
--    }
--
      *length = sizeof(USBD_BOSDesc);
      return (uint8_t *)USBD_BOSDesc;
  }
diff --cc lib/pbio/include/pbio/protocol.h
index 51b54b58,51b54b58..4fa00bca
--- a/lib/pbio/include/pbio/protocol.h
+++ b/lib/pbio/include/pbio/protocol.h
@@@ -301,9 -301,9 +301,7 @@@ extern const uint8_t pbio_pybricks_hub_
  
  extern const uint16_t pbio_gatt_device_info_service_uuid;
  extern const uint16_t pbio_gatt_firmware_version_char_uuid;
--extern const uint8_t pbio_gatt_firmware_version_char_uuid_128[];
  extern const uint16_t pbio_gatt_software_version_char_uuid;
--extern const uint8_t pbio_gatt_software_version_char_uuid_128[];
  extern const uint16_t pbio_gatt_pnp_id_char_uuid;
  
  extern const uint8_t pbio_lwp3_hub_service_uuid[];
diff --cc lib/pbio/src/protocol/pybricks.c
index e2305528,e2305528..6bf2ec9e
--- a/lib/pbio/src/protocol/pybricks.c
+++ b/lib/pbio/src/protocol/pybricks.c
@@@ -87,17 -87,17 +87,9 @@@ const uint16_t pbio_gatt_device_info_se
  
  /** Bluetooth Firmware Version Characteristic UUID. */
  const uint16_t pbio_gatt_firmware_version_char_uuid = 0x2A26;
--const uint8_t pbio_gatt_firmware_version_char_uuid_128[] = {
--    0x00, 0x00, 0x2A, 0x26, 0x00, 0x00, 0x10, 0x00,
--    0x80, 0x00, 0x00, 0x80, 0x5F, 0x9B, 0x34, 0xFB,
--};
  
  /** Bluetooth Software Version Characteristic UUID (Pybricks protocol version). */
  const uint16_t pbio_gatt_software_version_char_uuid = 0x2A28;
--const uint8_t pbio_gatt_software_version_char_uuid_128[] = {
--    0x00, 0x00, 0x2A, 0x28, 0x00, 0x00, 0x10, 0x00,
--    0x80, 0x00, 0x00, 0x80, 0x5F, 0x9B, 0x34, 0xFB,
--};
  
  /** Bluetooth PnP ID Characteristic UUID. */
  const uint16_t pbio_gatt_pnp_id_char_uuid = 0x2A50;
```